test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java

```diff
@@ -84,144 +84,237 @@ public static void main(String... args) throws Exception {
+                        Hello, Snippet!
+                    }
+                    """, null, "java"),
+                new SnippetAttributes("""
+                    {@snippet lang="properties" :
+                        Hello, Snippet!
+                    }
+                    """, null, "properties"),
+                new SnippetAttributes("""
+                    {@snippet lang="text" :
+                        Hello, Snippet!
+                    }
+                    """, null, "text"),
+                new SnippetAttributes("""
+                    {@snippet lang="" :
+                        Hello, Snippet!
+                    }
+                    """, null, null),
+                new SnippetAttributes("""
+                    {@snippet lang="foo" id="bar" :
+                        Hello, Snippet!
+                    }
+                    """, "bar", "foo")
         );
         ClassBuilder classBuilder = new ClassBuilder(tb, "pkg.A")
                 .setModifiers("public", "class");
         forEachNumbered(snippets, (s, i) -> {
             classBuilder.addMembers(
                     MethodBuilder.parse("public void case%s() { }".formatted(i))
-                            .setComments(s));
+                            .setComments("A method.", s.content()));
         });
         classBuilder.write(srcDir);
         javadoc("-d", outDir.toString(),
                 "-sourcepath", srcDir.toString(),
                 "pkg");
         checkExit(Exit.OK);
+        checkLinks();
         for (int j = 0; j < snippets.size(); j++) {
+            SnippetAttributes snippet = snippets.get(j);
             checkOutput("pkg/A.html", true,
                         """
                         <span class="element-name">case%s</span>()</div>
-                        <div class="block">
+                        <div class="block">A method.
+                         \s
                         <div class="snippet-container"><a href="#" class="snippet-copy" onclick="cop\
                         ySnippet(this)" aria-label="Copy" data-copied="Copied!"><img src="../copy.sv\
                         g" alt="Copy"></a>
-                        <pre class="snippet">
-                            Hello, Snippet!
-                        </pre>
+                        <pre class="snippet"%s><code%s>    Hello, Snippet!
+                        </code></pre>
                         </div>
-                        """.formatted(j));
+                        """.formatted(j, snippet.idAttribute(), snippet.langAttribute()));
         }
     }
 
+    /*
+     * Make sure the lang attribute is derived correctly from the snippet source file
+     * for external snippets when it is not defined in the snippet. Defining the lang
+     * attribute in the snippet should always override this mechanism.
+     */
+    @Test
+    public void testExternalImplicitAttributes(Path base) throws IOException {
+        Path srcDir = base.resolve("src");
```

Comment:
(Not a showstopper.) We need a test for hybrid snippets too. I will add it later.